sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1281,9 +1289,113 @@ CleanupEventListFromResetCmdList(std::vector<pi_event> &EventListToCleanup,
+
+      // Check that the comleted event has associated command list.
+      if (!(CompletedEvent->CommandList &&
+            CompletedEvent->CommandList.value() != Queue->CommandListMap.end()))
+        return PI_SUCCESS;
+
+      auto &CmdListEvents =
+          CompletedEvent->CommandList.value()->second.EventList;
+      auto CompletedEventIt =
+          std::find(CmdListEvents.begin(), CmdListEvents.end(), CompletedEvent);
+      if (CompletedEventIt != CmdListEvents.end()) {
+        // Remember the first event in this command list, we know that it
+        // is completed.
+        auto FirstEvent = CmdListEvents.front();
+        // We can cleanup all events prior to the completed event.
+        std::move(std::begin(CmdListEvents), CompletedEventIt + 1,
+                  std::back_inserter(EventListToCleanup));
+        CmdListEvents.erase(CmdListEvents.begin(), CompletedEventIt + 1);
+        // We can cleanup events from all command lists preceding the command
+        // list containing completed event.
+        for (auto &&It = Queue->CommandListMap.begin();
+             It != Queue->CommandListMap.end(); ++It) {
+          if (!It->second.EventList.empty()) {
+            auto LastEventInCmdList = It->second.EventList.back();
+            // If the command lists' last event is in the waitlist of the first
+            // event in the command list with completed event.
+            if (FirstEvent->WaitList.contains(LastEventInCmdList)) {
+              PI_CALL(Queue->resetCommandList(It, true, EventListToCleanup,
+                                              /* CheckStatus */ false));
+            }
+          }
+        }
+      }
+    } else {
+      // Fallback to resetCommandList over all command lists.
+      for (auto &&It = Queue->CommandListMap.begin();
+           It != Queue->CommandListMap.end(); ++It) {
+        PI_CALL(Queue->resetCommandList(It, true, EventListToCleanup,
+                                        /* CheckStatus */ true));
+      }
+    }
+  }
+  PI_CALL(CleanupEventListFromResetCmdList(EventListToCleanup, QueueLocked));
+  return PI_SUCCESS;
+}
+
+/// @brief Reset signalled command lists in the queue and put them to the cache
+/// of command lists. Also cleanup events associated with signalled command
+/// lists.
+/// @param Queue Queue where we look for signalled command lists and cleanup
+/// events.
+/// @param QueueLocked Indicates if the queue mutex is locked by caller.
+/// @param QueueSynced Hint informing that queue was synchronized before the
+/// call and no other commands were submitted after synchronization.
+/// @param CompletedEvent Hint providing an event which was synchronized before
+/// the call, in case of in-order queue it allows to cleanup all preceding
+/// events.
+/// @return PI_SUCCESS if successful, PI error code otherwise.
+static pi_result resetCommandLists(pi_queue Queue, bool QueueLocked = false,
+                                   bool QueueSynced = false,
```

Comment:
QueueLocked and QueueSynced have default value "false" but resetCommandLists is never called with either of them "true". The function CleanupEventsInImmCmdLists called from this function does use the QueueSynced flag, but I think these two parameters can be removed from this function.

Reply:
I agree, fixed.